vendor/k8s.io/kubernetes/cmd/kube-controller-manager/app/patch.go

```diff
@@ -1,12 +1,51 @@
 package app
 
 import (
+	"k8s.io/client-go/informers"
+	"k8s.io/client-go/rest"
 	"k8s.io/kubernetes/cmd/kube-controller-manager/app/options"
-	"k8s.io/kubernetes/pkg/controller"
 )
 
-// This allows overriding from inside the same process.  It's not pretty, but its fairly easy to maintain because conflicts are small.
-var CreateControllerContext func(s *options.CMServer, rootClientBuilder, clientBuilder controller.ControllerClientBuilder, stop <-chan struct{}) (ControllerContext, error) = createControllerContext
+var InformerFactoryOverride informers.SharedInformerFactory
 
-// StartInformers allows overriding inside of the same process.
-var StartInformers func(stop <-chan struct{}) = nil
+func ShimForOpenShift(controllerManager *options.CMServer, clientConfig *rest.Config) (func(), error) {
+	if len(controllerManager.OpenShiftConfig) == 0 {
+		return func() {}, nil
+	}
+
+	// TODO this gets removed when no longer take flags and no longer build a recycler template
```

Comment:
I've re-organized a bit and added comments about how we can get rid of each of these patches.